client/client.go

```diff
@@ -259,23 +268,35 @@ func (client *Client) Hijack(f func(*http.Request) (*http.Response, error)) {
 	client.doer = hijacked{f}
 }
 
+type doFlags struct {
+	NoTimeout bool
+}
+
 // do performs a request and decodes the resulting json into the given
 // value. It's low-level, for testing/experimenting only; you should
 // usually use a higher level interface that builds on this.
-func (client *Client) do(method, path string, query url.Values, headers map[string]string, body io.Reader, v interface{}) (statusCode int, err error) {
+func (client *Client) do(method, path string, query url.Values, headers map[string]string, body io.Reader, v interface{}, flags doFlags) (statusCode int, err error) {
 	retry := time.NewTicker(doRetry)
 	defer retry.Stop()
-	timeout := time.After(doTimeout)
+	timeout := time.NewTimer(doTimeout)
+	defer timeout.Stop()
+
 	var rsp *http.Response
+	var ctx context.Context = context.Background()
+	if !flags.NoTimeout {
+		cancelCtx, cancel := context.WithTimeout(ctx, doTimeout)
+		ctx = cancelCtx
+		defer cancel()
+	}
 	for {
-		rsp, err = client.raw(method, path, query, headers, body)
+		rsp, err = client.raw(ctx, method, path, query, headers, body)
 		if err == nil || method != "GET" {
 			break
 		}
 		select {
 		case <-retry.C:
 			continue
-		case <-timeout:
+		case <-timeout.C:
```

Comment:
having this timeout when one set flags.NoTimeout seems unexpected, to me

Reply:
The `timeout.C` imposes an upper time limit on the retry loop. The retry loop itself is AFAIK is for graceful handling of scenarios where snapd is being restarted. However, none of these fix the problem of `client.raw()` getting stuck, which is addressed by the WithTimeout() context passed with the request.